### PR TITLE
Add keep stage to LogQL

### PR DESCRIPTION
**Pull request by salvacorts**
**What this PR does / why we need it**:
This PR adds a new `keep` stage to LogQL. It allows filtering specific labels and helps to avoid hitting the _max streams per query_ limit by keeping only specific labels in the label set. It works the opposite of [drop stage](https://grafana.com/docs/loki/latest/logql/log_queries/#drop-labels-expression).

**Special notes for your reviewer**:
We do not drop [special error labels](https://github.com/grafana/loki/blob/edc6b0bff76714ff584c20d7ff0235461a4f4a88/pkg/logqlmodel/error.go#L18-L20) added by Loki at query time. Users will have to explicitly drop them by using the drop stage.

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

### CODE REVIEW

File: pkg/logql/log/keep_labels.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	return line, true
+}
+
+func (kl *KeepLabels) RequiredLabelNames() []string {
+	return []string{}
+}
+
+func isSpecialLabel(lblName string) bool {
+	switch lblName {
+	case logqlmodel.ErrorLabel, logqlmodel.ErrorDetailsLabel, logqlmodel.PreserveErrorLabel:
```
Comment (sandeepsukhani) [inline pkg/logql/log/keep_labels.go:67]: Although ErrorLabel and ErrorDetailsLabel are not part of the label set and are stored separately, I have added a check for them here to keep it future-proof.

**Comment by salvacorts** [comment]
Thanks for adding the docs and the error check, Sandeep. I think we can mark this PR as ready :) wdyt?

File: docs/sources/query/log_queries/_index.md [hunk 2]
```diff
@@ -641,3 +641,42 @@ the result will be
...
 {app="other-service", host="grafana.net", job="varlogs", method="GET", status="200"} {"app": "other-service", "level": "info", "method": "GET", "path": "/", "host": "grafana.net", "status": "200"}
 ```
+
+### Keep Labels expression
+
+**Syntax**:  `|keep name, other_name, some_name="some_value"`
+
+The `| keep` expression will keep only the specified labels in the pipeline and drop all the other labels.
+
+**NOTE**: keep stage will not drop __error__ or __error_details__ labels added by Loki at query time. To drop these labels, please refer to [drop](#drop-labels-expression) stage.
```
Comment (JStickler) [inline docs/sources/query/log_queries/_index.md:651]:
```suggestion
{{% admonition type="note" %}}
The keep stage will not drop the  __error__ or __error_details__ labels added by Loki at query time. To drop these labels, refer to [drop](#drop-labels-expression) stage.
{{% /admonition %}}
```

File: docs/sources/query/log_queries/_index.md [hunk 3]
```diff
@@ -641,3 +641,42 @@ the result will be
...
+
+Query examples:
+
+For the query `{job="varlogs"}|json|keep level, method="GET"`, with below log line
```
Comment (JStickler) [inline docs/sources/query/log_queries/_index.md:655]:
```suggestion
For the query `{job="varlogs"}|json|keep level, method="GET"`, with the following log lines:
```

File: docs/sources/query/log_queries/_index.md [hunk 4]
```diff
@@ -641,3 +641,42 @@ the result will be
...
+```
+
+the result will be
+
+```
+{level="info", method="GET"} {"level": "info", "method": "GET", "path": "/", "host": "grafana.net", "status": "200"}
+{level="info"} {"level": "info", "method": "POST", "path": "/", "host": "grafana.net", "status": "200"}
+```
+
+For the query `{job="varlogs"}|json|keep level, tenant, app=~"some-api.*"`, with below log lines
```
Comment (JStickler) [inline docs/sources/query/log_queries/_index.md:669]:
```suggestion
For the query `{job="varlogs"}|json|keep level, tenant, app=~"some-api.*"`, with the following log lines:
```

**Review by JStickler — COMMENTED** [review]
[docs team]

**Review by JStickler — APPROVED** [review]
[docs team] LGTM

**Pull request merged**